crates/iota-data-ingestion/src/workers/kv_store.rs

```diff
@@ -152,32 +144,53 @@ impl KVStoreWorker {
         Ok(())
     }
 
-    async fn upload_blob<V: Serialize + std::marker::Send>(
+    /// Uploads checkpoint contents to storage, with automatic fallback from
+    /// DynamoDB to S3.
+    ///
+    /// This function attempts to store checkpoint contents in DynamoDB first.
+    /// If that fails (typically due to size constraints), it automatically
+    /// falls back to uploading the contents to S3 instead.
+    async fn upload_checkpoint_contents<V: Serialize + std::marker::Send>(
         &self,
         key: Vec<u8>,
         value: V,
     ) -> anyhow::Result<()> {
-        let body = bcs::to_bytes(value.borrow())?.into();
-        self.s3_client
-            .put_object()
-            .bucket(self.bucket_name.clone())
-            .key(base64_url::encode(&key))
-            .body(body)
+        let bcs_bytes = bcs::to_bytes(value.borrow())?;
+
+        let attributes = HashMap::from([
+            (
+                "digest".to_string(),
+                AttributeValue::B(Blob::new(key.clone())),
+            ),
+            (
+                "type".to_string(),
+                AttributeValue::S(ItemType::CheckpointContents.to_string()),
+            ),
+            (
+                "bcs".to_string(),
+                AttributeValue::B(Blob::new(bcs_bytes.clone())),
+            ),
+        ]);
+
+        let res = self
+            .dynamo_client
+            .put_item()
+            .table_name(&self.table_name)
+            .set_item(Some(attributes))
             .send()
-            .await?;
-        Ok(())
-    }
+            .await;
 
-    fn type_name(table: KVTable) -> String {
-        match table {
-            KVTable::Transactions => "tx",
-            KVTable::Effects => "fx",
-            KVTable::EventsByTxDigest => "evtx",
-            KVTable::Objects => "ob",
-            KVTable::CheckpointSummary => "cs",
-            KVTable::TransactionToCheckpoint => "tx2c",
+        if res.is_err() {
```

Comment:
Would it make sense to log the dynamodb error here? Or it's not that relevant?

Reply:
[fixed](https://github.com/iotaledger/iota/pull/5855/commits/619ddc1f611b8278e421903fd8f20d4efffedaac)